lib/eco/autolboxdetector.py

```diff
@@ -63,11 +64,16 @@ def find_terminal(self, node, version=None):
         return node.next_term
 
     def detect_lbox(self, errornode):
-        # Try heuristic #2 first; if it fails to find an automatic language
-        # box, try heuristic #1
-        valid = self.detect_lbox_h2(errornode)
-        if not valid:
-            valid = self.detect_lbox_h1(errornode)
+        # Try history based heuristic first; if it fails to find an automatic
+        # language box, try the stack based one; if that fails, try the
+        # line-based heuristic.
+        valid = []
+        if config.AUTOLBOX_HEURISTIC_HIST:
+            valid = self.heuristic_history(errornode)
+        if not valid and config.AUTOLBOX_HEURISTIC_STACK:
```

Comment:
An obvious question is what happens if we remove the `not valid and` here (and remove duplicates)?

Reply:
Ah, OK, this is addressed in the following commit.